Add a check for receiver to validate minimum required fees

jbesraa opened the pull request:
resolves #105 

    Payjoin receiver can set minimum `fee_rate`
    in order to reject payjoin requests that contain
    a psbt which doesnt meet the minimum `fee_rate`
    threshold.

    We add this functionality to `check_broadcast_suitability`
    function which previously was called `check_can_broadcast`.

jbesraa commented on payjoin/src/receive/mod.rs:
```diff
@@ -347,6 +348,19 @@ impl UncheckedProposal {
+    /// Check that the Original PSBT meet the minimum feerate receiver requires.
+    pub fn check_if_meet_min_feerate(self, receiver_min_feerate: FeeRate) -> Result<(), Error> {
+        let sender_fee_rate = self.params.min_feerate;
+        // let sender_addtional_fee_contribution = self.params.additional_fee_contribution;
```
what else should we take into account in this check?
we have another couple of params that can be interesting:
1) `self.psbt.fee()` (this thing returns `Bitcoin::Amount` but I guess we can convert it to `bitcoin::FeeRate` 
2) `self.params.addition_fee_contr` - this need to be validated(other index input should be set) first which we do in a different check

DanGould replied:
`params.min_feerate` != original psbt feerate. That parameter can be set arbitrarily. What we want to check is the effective FeeRate of `self.psbt` to make sure this is a reliable fallback if the sender disappears. Otherwise, receiver would be subject to probing attacks where the original psbt fee rate is too low for the receiver to actually ever get it included in a block in the fallback case. As you mention, `additional_fee_contribution` is handled later on and can be ignored in this check.

Perhaps instead of `check_if_meet_min_feerate` this functionality could be combined with `check_can_broadcast`. Since that is already a `testmempoolaccept` check to prevent probing, which is the basis for this issue.

Something like `check_broadcast_suitability(self, min_feerate: Option<bitcoin::FeeRate>, can_broadcast: impl Fn(&bitcoin::Transaction) -> Result<bool, Error>` could replace check_can_broadcast.

min_feerate could be enforced on `assume_interactive_receiver(self, min_feerate: Option<bitcoin::FeeRate>)`` using a shared helper function, too

As a bonus this uses our typestate machine so the check can't be so easily overlooked.

jbesraa replied:
sorry for the delay here, working on it

jbesraa replied:
thanks for the patience
I pushed a new version and split the commits.

jbesraa commented on payjoin/src/receive/mod.rs:
```diff
@@ -850,16 +866,17 @@ mod test {
-        )
+        let min_feerate_param = format!("minfeerate={}", min_feerate.unwrap_or(0));
+        let query = format!(
+            "maxadditionalfeecontribution=182&additionalfeeoutputindex=0&{}",
```
this previous query used `?` to separate between pairs, which is not correct. is should be `&` (refer to `url_fromencoded` crate where they split by `&`)

DanGould replied:
Nice catch. This should be its own commit

jbesraa commented on payjoin/src/receive/mod.rs:
```diff
@@ -295,8 +295,9 @@ pub trait Headers {
-/// that the transaction can be broadcast with check_can_broadcast. Otherwise it is safe to
+/// that the transaction can be broadcast with check_broadcast_suitability. Otherwise it is safe to
 /// call assume_interactive_receive to proceed with validation.
+#[derive(Debug, Clone)]
```
Added for testing/debugging purposes

jbesraa commented on payjoin/src/receive/mod.rs:
```diff
@@ -359,10 +360,18 @@ impl UncheckedProposal {
     ///
     /// Call this after checking downstream.
-    pub fn check_can_broadcast(
+    pub fn check_broadcast_suitability(
```
we probably want to update the documentation of the function to reflect the current change

jbesraa commented on payjoin/src/receive/optional_parameters.rs:
```diff
@@ -4,7 +4,7 @@ use std::fmt;
 use log::warn;
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Added for testing/debugging purposes

DanGould commented on payjoin/src/receive/error.rs:
```diff
@@ -65,6 +65,10 @@ pub(crate) enum InternalRequestError {
+    /// Fee rate provided by sender is too low
+    /// First argument is the fee rate provided by the sender
+    /// Second argument is the minimum fee rate required by the receiver
+    FeeTooLow(bitcoin::Amount, bitcoin::FeeRate),
```
Not sure I understand. Absolute fee and feerate are distinct, no?

jbesraa replied:
yup. you are right. took a second to click

jbesraa commented on payjoin/src/receive/mod.rs:
```diff
@@ -841,7 +841,9 @@ mod test {
     }
 
-    fn proposal_from_test_vector() -> Result<UncheckedProposal, RequestError> {
+    fn proposal_from_test_vector(
+        min_feerate: Option<u64>,
+    ) -> Result<UncheckedProposal, RequestError> {
```
its used to construct the query a few lines below

jbesraa commented on payjoin/src/receive/mod.rs:
```diff
@@ -850,16 +852,17 @@ mod test {
         let headers = MockHeaders::new(body.len() as u64);
-        UncheckedProposal::from_request(
-            body,
-            "?maxadditionalfeecontribution=182?additionalfeeoutputindex=0",
```
its not really needed here.. also not expected by `from_urlencoded`

jbesraa commented on payjoin/src/receive/mod.rs:
```diff
@@ -852,7 +873,7 @@ mod test {
         UncheckedProposal::from_request(
             body,
-            "?maxadditionalfeecontribution=182?additionalfeeoutputindex=0",
+            "maxadditionalfeecontribution=182&additionalfeeoutputindex=0&minfeerate=0",
```
not required

jbesraa commented:
@DanGould 
Rebased from master branch which required some minor changes to `v2` stuff, and removed not required changes in the scope of this pull request.
We talked about test vectors for different PSBTs, but I was not able to find a quick solution to generate PSBTs with some configuration regrading fee or weights, might be something we want to develop internally.
I think this pr can go on and we can look into more robust testing in the (near) future.

DanGould commented on payjoin/src/receive/mod.rs:
```diff
@@ -345,26 +346,46 @@ impl UncheckedProposal {
     /// The Sender's Original PSBT
-    pub fn extract_tx_to_schedule_broadcast(&self) -> bitcoin::Transaction {
-        self.psbt.clone().extract_tx()
+    pub fn psbt_transaction(&self) -> bitcoin::Transaction { self.psbt.clone().extract_tx() }
```
This name change seems unrelated to the receiver check this PR is about and I'd prefer it wasn't made. extract_tx (for the specific api reason) tries to convey why the api is available, and extract lets you know it's calculating rather than returning something from memory.

Updating this would also merit a change to the documentation at the front matter of this file

DanGould commented on payjoin/src/receive/mod.rs:
```diff
@@ -345,26 +346,46 @@ impl UncheckedProposal {
+    pub fn psbt_transaction(&self) -> bitcoin::Transaction { self.psbt.clone().extract_tx() }
+
+    /// The Sender's Original PSBT fee rate.
+    pub fn psbt_fee_rate(&self) -> Result<FeeRate, Error> {
```
Not sure this function should be public, especially since it returns BadRequest, which doesn't make much sense as a return type for a fee_rate here

DanGould commented on payjoin/src/receive/mod.rs:
```diff
@@ -345,26 +346,46 @@ impl UncheckedProposal {
+        let original_psbt_fee_rate = self.psbt_fee_rate()?;
+        if let Some(min_fee_rate) = min_fee_rate {
+            if original_psbt_fee_rate < min_fee_rate {
+                return Err(Error::BadRequest(
+                    InternalRequestError::PsbtBelowFeeRate(original_psbt_fee_rate, min_fee_rate)
```
I suggest implicit conversion to Error by using `.into()` on its own. See `return Err(InternalRequestError::InvalidContentType(content_type.to_owned()).into());`

DanGould commented on payjoin/src/receive/mod.rs:
```diff
@@ -873,6 +894,7 @@ mod test {
         use bitcoin::{Address, Network};
 
         let proposal = proposal_from_test_vector().unwrap();
+        assert_eq!(proposal.psbt_fee_rate().unwrap().to_sat_per_vb_floor(), 2);
```
I understand why you might make psbt_fee_rate public for this function, but it's so brief and nonessential to the payjoin api I'd rather see it repeated in the test than exposed publically.

jbesraa replied:
the pub was a leftover from when i was testing in integration, removed that. 
here in the unit test it can access all the functions/structs in the same module so its fine.

DanGould commented on payjoin/src/receive/v2.rs:
```diff
@@ -323,10 +323,10 @@ impl Enrolled {
 /// call assume_interactive_receive to proceed with validation.
 pub struct UncheckedProposal {
-    inner: super::UncheckedProposal,
+    pub inner: super::UncheckedProposal,
```
Why make this public? Ideally these typestate machines are kept opaque so that the only modifications happen through the machine.

DanGould commented on payjoin/src/receive/v2.rs:
```diff
@@ -364,11 +364,11 @@ impl UncheckedProposal {
     /// Call this after checking downstream.
-    pub fn check_can_broadcast(
+    pub fn check_broadcast_suitability(
         self,
         can_broadcast: impl Fn(&bitcoin::Transaction) -> Result<bool, Error>,
     ) -> Result<MaybeInputsOwned, Error> {
-        let inner = self.inner.check_can_broadcast(can_broadcast)?;
+        let inner = self.inner.check_broadcast_suitability(None, can_broadcast)?;
```
This v2 UncheckedProposal should have the same interface for `self.inner.check_boradcast_suitablity` and accept `min_feerate`. Hardcoding`None` should raise a yellow flag.

DanGould commented on payjoin/src/receive/error.rs:
```diff
@@ -150,6 +157,17 @@ impl fmt::Display for RequestError {
             InternalRequestError::ParsePsbt(e) => write_error(f, "Error parsing PSBT:", e),
             #[cfg(feature = "v2")]
             InternalRequestError::Utf8(e) => write_error(f, "Error parsing PSBT:", e),
+            InternalRequestError::PsbtBelowFeeRate(
+                original_psbt_fee_rate,
+                receiver_min_fee_rate,
+            ) => write_error(
+                f,
+                "original-psbt-rejected",
+                &format!(
+                    "Original PSBT fee rate too low: {} < {}.",
+                    original_psbt_fee_rate, receiver_min_fee_rate
+                ),
+            ),
```
Very nice! 🤌

DanGould commented on payjoin/src/receive/error.rs:
```diff
@@ -86,6 +86,13 @@ pub(crate) enum InternalRequestError {
     Utf8(std::string::FromUtf8Error),
+    /// Original PSBT fee rate is below minimum fee rate set by the receiver.
+    ///
+    /// First argument is the calculated fee rate of the original PSBT.
+    /// [`UncheckedProposal::psbt_fee_rate`]
```
remember to remove this comment if choosing to remove psbt_fee_rate from the public interface.

DanGould reviewed (changes requested):
Getting much closer 🔥 I think this gets the job done and only request a few cleanups

jbesraa commented:
Thanks for the input @DanGould 
I have addressed all the above

DanGould commented:
Ci failed to catch the need to change here: https://github.com/payjoin/rust-payjoin/blob/624ab7f0f339d0232644e507b9b66925f7ecfd32/payjoin-cli/src/app.rs#L542

We need to make sure payjoin-cli builds in v2
